File: competitive-programming/leet-code/464/canIWin.js
var canIWin = function (maxn, total) {
    if (total === 0) return true;

    let sumn = (maxn * (maxn + 1)) / 2;
    if (sumn < total) return false;

    let masks = new Array(1 << maxn);
    let bitMaskmaxn = (1 << maxn) - 1;

    function solve(currentMask, total) {
        if (currentMask <= 0) return false;
        if (total <= 0) return false;
        if (masks[currentMask] !== undefined) return masks[currentMask];

        for (let n = 1, bitMask = 1; n <= maxn && bitMask <= currentMask; ++n, bitMask = bitMask << 1) {
            let nNotTaken = currentMask & bitMask;
            if (nNotTaken) {
                let currentMaskWithoutN = currentMask ^ bitMask;
                let totalWithoutN = total - n;
                if (!solve(currentMaskWithoutN, totalWithoutN)) {
                    return (masks[currentMask] = true);
                }
            }
        }
        return (masks[currentMask] = false);
    }

    return solve(bitMaskmaxn, total);
};

File: competitive-programming/leet-code/464/can_i_win.py
def can_i_win(maxn: int, total: int) -> bool:
    if total == 0:
        return True
    
    sumn = (maxn * (maxn + 1)) // 2
    if sumn < total:
        return False
    
    masks = [None] * (1 << maxn)
    bit_maskmaxn = (1 << maxn) - 1
    
    def solve(current_mask: int, total: int) -> bool:
        if current_mask <= 0:
            return False
        if total <= 0:
            return False
        if masks[current_mask] is not None:
            return masks[current_mask]
        
        n = 1
        bit_mask = 1
        while n <= maxn and bit_mask <= current_mask:
            nNotTaken = current_mask & bit_mask
            if nNotTaken:
                current_mask_without_n = current_mask ^ bit_mask
                total_without_n = total - n
                if not solve(current_mask_without_n, total_without_n):
                    masks[current_mask] = True
                    return True
            n += 1
            bit_mask = bit_mask << 1
        
        masks[current_mask] = False
        return False
    
    return solve(bit_maskmaxn, total)